tests/hypervolume_tests/test_box_decomposition.py

```diff
@@ -0,0 +1,139 @@
+
+
+_EPS = 1e-12
+
+
+class InstanceGenerator(Protocol):
+    def __call__(self, n_trials: int, n_objectives: int, seed: int) -> np.ndarray:
+        raise NotImplementedError
+
+
+def _extract_pareto_sols(loss_values: np.ndarray) -> np.ndarray:
+    sorted_loss_vals = np.unique(loss_values, axis=0)
+    on_front = _is_pareto_front(sorted_loss_vals, assume_unique_lexsorted=True)
+    return sorted_loss_vals[on_front]
+
+
+def _generate_uniform_samples(n_trials: int, n_objectives: int, seed: int) -> np.ndarray:
+    rng = np.random.RandomState(seed)
+    return np.maximum(_EPS, rng.random((n_trials, n_objectives)))
+
+
+def _generate_instances_with_negative(n_objectives: int, n_trials: int, seed: int) -> np.ndarray:
+    rng = np.random.RandomState(seed)
+    return _extract_pareto_sols(rng.normal(size=(n_trials, n_objectives)))
+
+
+def _generate_concave_instances(n_trials: int, n_objectives: int, seed: int) -> np.ndarray:
+    """See Section 4.2 of https://arxiv.org/pdf/1510.01963"""
+    points = _generate_uniform_samples(n_trials, n_objectives, seed)
+    loss_values = points / np.maximum(_EPS, np.sum(points**2, axis=-1))[:, np.newaxis]
+    return _extract_pareto_sols(loss_values)
+
+
+def _generate_convex_instances(n_trials: int, n_objectives: int, seed: int) -> np.ndarray:
+    """See Section 4.2 of https://arxiv.org/pdf/1510.01963"""
+    points = _generate_uniform_samples(n_trials, n_objectives, seed)
+    loss_values = 1 - points / np.maximum(_EPS, np.sum(points**2, axis=-1))[:, np.newaxis]
+    return _extract_pareto_sols(loss_values)
+
+
+def _generate_linear_instances(n_trials: int, n_objectives: int, seed: int) -> np.ndarray:
+    """See Section 4.2 of https://arxiv.org/pdf/1510.01963"""
+    points = _generate_uniform_samples(n_trials, n_objectives, seed)
+    loss_values = points / np.maximum(_EPS, np.sum(points, axis=-1))[:, np.newaxis]
+    return _extract_pareto_sols(loss_values)
+
+
+def _calculate_hypervolume_improvement(
+    lbs: np.ndarray, ubs: np.ndarray, new_points: np.ndarray
+) -> np.ndarray:
+    diff = np.maximum(0.0, ubs - np.maximum(new_points[..., np.newaxis, :], lbs))
+    # The minimization version of Eq. (1) in https://arxiv.org/pdf/2006.05078.
+    return np.sum(np.prod(diff, axis=-1), axis=-1)
+
+
+def _verify_exact_hypervolume_improvement(pareto_sols: np.ndarray) -> None:
+    ref_point = np.max(pareto_sols, axis=0)
+    ref_point = np.maximum(ref_point * 1.1, ref_point * 0.9)
+    n_sols = pareto_sols.shape[0]
+    # LOO means leave one out.
+    loo_mat = ~np.eye(n_sols, dtype=bool)
+
+    hv = compute_hypervolume(pareto_sols, ref_point, assume_pareto=True)
+    correct = hv - np.array([compute_hypervolume(pareto_sols[loo], ref_point) for loo in loo_mat])
+    ans = np.empty_like(correct)
+    for i, loo in enumerate(loo_mat):
+        lbs, ubs = get_non_dominated_box_bounds(pareto_sols[loo], ref_point)
+        ans[i] = _calculate_hypervolume_improvement(lbs, ubs, pareto_sols[np.newaxis, i])[0]
```

Comment:
This logic looks too complex. How about just calculating the hypervolume by WFG(compute_hypervolume) and the box decomposition and validating they are close?